### ISSUE TITLE
Assembly V2 all Lepto raw assemblies

### ISSUE THREAD

**Issue by rx32940** [rx32940/Lepto-Phylogeography#4]
- last version of the raw reads for Leptospira assemblies **did not have quality check** nor **trimming** done before assembly.
- to assure the quality of all isolates before assemblies, we will **redo** assemblies for all raw reads with the **full work flow**
---
- [ ] fastqc
- [ ] trimming
- [ ] spades

**Comment by rx32940** [comment]
## FastQC with 51 raw reads records with the collection dates

- some of these isolates have multiple runs (technical replicates) 
   - the runs for the same isolates were merged with ```cat``` and named by  ${biosample_accession}.fastq.gz
- some with the reverse reads only with one base pair long
   - dropping reverse reads, treat as single end, rename by  ${biosample_accession}.fastq.gz
- some were pair end reads
   - fastq forward and reverse file separately, rename by  ${biosample_accession}_1(2).fastq.gz

- **list of single/pair-end** raw reads record can be find in:
```
/scratch/rx32940/merged_runs/single(pair).txt
```
---

- complete information for the sequences of these isolates can be find in:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/seq_info/dated51_biosample_sra.xlsx
```

- all the raw reads fastq files can be find here:
```
/scratch/rx32940/merged_runs/single(pair)
```
* **note**: SRR10301832 is single read

- the FastQC report and aggregated report by multiQC is store in the dir:
```
/scratch/rx32940/Lepto_Work/fastqc
```
- interactive qualities of these sequences are in:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/fastqc
```

[pipeline to obtain the results above](https://github.com/rx32940/Lepto-Phylogeography/blob/f1ca10238f4462f0f9d0d3e9d7ab898ee1f07ae0/dated_assemblies_pipeline.sh)

- from the fastqc report above, all sequence but **SAMN13046976** has passed the quality check
   - SAMN13046976 was sequenced by Ion torrant and was found already with assemblies available on NCBI, thus we will use the assemblies downloaded:
```
/scratch/rx32940/merged_runs/SAMN13046976.fsa_nt
```

**Comment by rx32940** [comment]
## Use Trimmomatic to trim the adapters and bad quality reads from raw reads

- trimming was done for single-end samples and pair-end samples separately
- the results can be find in the dir:
```
/scratch/rx32940/Lepto_Work/trim/single(pair)
```

### For **single-end** records:

- ILLUMINACLIP: cut off Illumina adapters _TruSeq3-SE.fa:2:30:10_
- LEADING: Cut bases off the start of a read, if below a threshold quality, _3_
- LEADING: Cut bases off the start of a read, if below a threshold quality, _3_
- SLIDINGWINDOW: Perform a sliding window of length _4_ trimming, cutting once the average quality,_15_, within the window falls below a threshold.
- MINLEN: Drop the read if it is below a specified length, _36_
- OUTPUT: **${biosample}_trimmed.fastq.gz**

### For **pair-end** records:

- ILLUMINACLIP: cut off Illumina adapters _TruSeq3-PE.fa:2:30:10_
- LEADING: Cut bases off the start of a read, if below a threshold quality, _3_
- LEADING: Cut bases off the start of a read, if below a threshold quality, _3_
- SLIDINGWINDOW: Perform a sliding window of length _4_ trimming, cutting once the average quality,_15_, within the window falls below a threshold.
- MINLEN: Drop the read if it is below a specified length, _36_
- OUTPUT: 
**${biosample}_1_paired_trimmed.fastq.gz ${biosample}_1_unpaired_trimmed.fastq.gz 
${biosample}_2_paired_trimmed.fastq.gz ${biosample}_2_unpaired_trimmed.fastq.gz**

* unpaired output are reads with their paired reads from the opposite end dropped due to bad quality. Can be used as single-end reads during assembly

[pipeline to obtain the results above](https://github.com/rx32940/Lepto-Phylogeography/blob/3839f6684b9886ef4fdc6065e36411e22bdc71e0/dated_assemblies_pipeline.sh)

**Comment by rx32940** [comment]
## Use Trimmed sequence for assembly with SPADES

- single and pair-end sequences were assembled separately
- unpaired trimmed reads (pair reads cut due to bad quality were treated as single-end reads

---
- Raw reads from **St. Kitts**  for Dr. Rajeev's work recently has their assemblies uploaded for their raw reads
- [x] compare the QUAST results between their uploaded assemblies with mine on the same isolate

[pipeline for assembly](https://github.com/rx32940/Lepto-Phylogeography/blob/6b35b5a70730c05af4ed24a4135db8f38a2beb6a/dated_assemblies_pipeline.sh)

**Comment by rx32940** [comment]
## Use QUAST assemblies evaluation 

- isolates from the 50 raw reads records with collection date (excluding SAMN13046976) were QUASTed with the reference genome from their corresponding species (only _L.Interrogans_ and _L.borgpetersenii_ for this set of isolates)
- reference genome can be find in the dir:
```
/scratch/rx32940/reference
```
[pipeline for QUAST](https://github.com/rx32940/Lepto-Phylogeography/blob/4d5d68b5a1e40b9cf9ea2f910bd9161bca73acc6/dated_assemblies_pipeline.sh)

**Comment by rx32940** [comment]
> * Raw reads from **St. Kitts**  for Dr. Rajeev's work recently has their assemblies uploaded for their raw reads
> * [x]  compare the QUAST results between their uploaded assemblies with mine on the same isolate

- assemblies uploaded to **NCBI by Alejandro** (PRJNA543109) are way more complete and better quality than my draft assembled genomes
- Taking suggestions from **Xueting**, I will compare the qualities (QUAST reports) of the assemblies downloaded directly from NCBI and my own assemblies from raw reads **separately**.

_" 3) I hope you can compare the QUAST reports separately for your own assembly and the samples that have been assembled in the database before you have an overall QC analysis. This way, you can have a better understanding of your data, and how reasonable you can use the data from different approaches. And you can also come up with ways to conduct sensitivity analysis later. "_ 

- this leaves only **32 self-assembled genomes** in raw reads with the collection dates
---

- 18 assemblies by Alejandro (PRJNA543109) were downloaded and can be found in the dir:
```
/scratch/rx32940/All_Lepto_Assemblies/sree_18/assemblies
```
[code for this QUAST job](https://github.com/rx32940/Lepto-Phylogeography/blob/34db9b30b6fa4a23d474e58dd11607c65f7aab35/dated_assemblies_pipeline.sh)

**Comment by lsalvador** [comment]
> * last version of the raw reads for Leptospira assemblies **did not have quality check** nor **trimming** done before assembly.
> * to assure the quality of all isolates before assemblies, we will **redo** assemblies for all raw reads with the **full work flow**
> 
> * [ ]  fastqc
> * [ ]  trimming
> * [ ]  spades

test

**Comment by lsalvador** [comment]
https://github.com/rx32940/Lepto-Phylogeography/issues/4#issue-567197826

test 3

**Comment by rx32940** [comment]
Read these two articles for bacteria genome analysis (in Mendeley):
- Swain, M. T., Tsai, I. J., Assefa, S. A., Newbold, C., Berriman, M., & Otto, T. D. (2012). A post-assembly genome-improvement toolkit (PAGIT) to obtain annotated genomes from contigs. Nature Protocols, 7(7), 1260–1284. https://doi.org/10.1038/nprot.2012.068

- Ahrenfeldt, J., Skaarup, C., Hasman, H., Pedersen, A. G., Aarestrup, F. M., & Lund, O. (2017). Bacterial whole genome-based phylogeny: construction of a new benchmarking dataset and assessment of some existing methods. BMC Genomics, 18(1), 19. https://doi.org/10.1186/s12864-016-3407-6

- also review Xu et al. (2016) and Fouts et al. (2016) for analysis done on bacterial genomes

**Comment by rx32940** [comment]
### Reference folder updated 

- now with 45 different species from rest_211,picardeau_313 and dated_32 assemblies
- species for each isolate can be extracted by:
   - need to get xml file from NCBI with a list of biosamples using batch entrez
   - get the biosample with [extract_xml_metadata.ipynb](https://github.com/rx32940/Lepto-Phylogeography/blob/d354e943fd3c72a5d2214ba3ce039a1245991d7a/extract_xml_metadata.ipynb) and also the unique species 
    - upload species_dict.csv to sapelo2
    - change path to the dict file in [python script](https://github.com/rx32940/Lepto-Phylogeography/blob/d354e943fd3c72a5d2214ba3ce039a1245991d7a/get_biosample_species.py) 

- all three datasets were quasted with their corresponding species' reference genomes using their pipelines with species extracted from python script embedded inside their assembly pipelines.
   - all species with species name _Leptospira sp._ was replaced with _Leptospira interrogans_ (in *_replaced.csv dict file)
   - all isolates in dated_32 are _Leptospira interrogans_

**Comment by rx32940** [comment]
- find QUAST report for all three datasets in:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/
```


### Rest_211 dataset

- after spades and Quast, only **207 assemblies** left
   - SAMEA864154, too little bp, failed to assemble
   - SAMEA864155, too little bp, failed to assemble
   - SAMEA864115, no trimming, didn't assemble
  - SAMEA864093, fail to assemble

assemblies Dir:
```
/scratch/rx32940/Lepto_Work/rest_211/assemblies/$biosample_acc/scaffolds.fasta
```
 
pipeline to the Quast step:
[rest_assembly_pipeline.sh](https://github.com/rx32940/Lepto-Phylogeography/blob/d354e943fd3c72a5d2214ba3ce039a1245991d7a/rest_assembly_pipeline.sh)

- [ ] filter out assemblies with bad qualities
   number of assemblies left:

### picardeau_313 dataset

- spades and Quast have done for all 313 assemblies, dir:

```
/scratch/rx32940/Lepto_Work/picardeau_313/assemblies/$biosample_acc/scaffolds.fasta
```

pipeline to the Quast step:
[picardeau_assemble_pipeline.sh](https://github.com/rx32940/Lepto-Phylogeography/blob/d354e943fd3c72a5d2214ba3ce039a1245991d7a/picardeau_assemble_pipeline.sh)

- [ ] filter out assemblies with bad qualities
   number of assemblies left:

### dated_32 dataset

- assemblied all 32 assemblies dir:
```
/scratch/rx32940/Lepto_Work/dated_32/assemblies/$biosample_acc/scaffolds.fasta
```

pipeline to the Quast step:
[dated_assemblies_pipeline.sh](https://github.com/rx32940/Lepto-Phylogeography/blob/d354e943fd3c72a5d2214ba3ce039a1245991d7a/dated_assemblies_pipeline.sh)

- [ ] filter out assemblies with bad qualities
   number of assemblies left:

**Comment by rx32940** [comment]
# Isolates Low Percentage of bp Aligned to Reference

**genome fraction (QUAST):** The total number of aligned bases in the reference divided by the genome size

- both **Picardeau_313** and **rest_207** have isolates have shown <10% genome fraction with their reported corresponding species' reference genome. 

- to assure the quality of these assemblies, I compared their reported species with the species analyzed by The NCBI **SRA Taxonomy Analysis Tool (STAT)**, which calculates the taxonomic distribution of reads from SRA records and percentage of reads under the taxonomy.
   
- **"_Leptopsira sp._"**: this can mean either the composition of the distribution was shared by multiple species of Leptospira without a dominant species or STAT only classified reads as the _Leptospira_ genus as dominant composition without classifying a dominant species.   

- I classified isolates with **STAT percentage** less than **70%** as contaminated samples, to not affect the results of the analysis, these isolates will be **removed** from downstream analysis 

### Picardeau Low Genome Fraction Isolates

picardeau | reported species | NCBI blast analysis | percentage | keep?
-- | -- | -- | -- | --
SAMEA5168064 | Leptospira sp. | Leptospira sp. | 28.4 |  
SAMEA5168065 | Leptospira sp. | Leptospira sp. | 29.7 |  
SAMEA5168073 | Leptospira sp. | Leptospira sp. | 26.1 |  
SAMEA5168081 | Leptospira biflexa | Leptospira yanagawae | 94 | yes
SAMEA5168084 | Leptospira biflexa | Leptospira sp. | 18.8 |  
SAMEA5168090 | Leptospira meyeri | Leptospira biflexa | 98.7 | yes
SAMEA5168097 | Leptospira biflexa | Leptospira sp. | 24 |  
SAMEA5168098 | Leptospira biflexa | Leptospira sp. | 26.8 |  
SAMEA5168099 | Leptospira biflexa | Leptospira sp. | 88.2 | yes
SAMEA5168113 | Leptospira borgpetersenii | Leptospira mayottensis | 99 | yes
SAMEA5168128 | Leptospira biflexa | Leptospira sp. | 32.3 |  
SAMEA5168129 | Leptospira biflexa | Leptospira sp. | 99.2 | yes
SAMEA5168130 | Leptospira venezuelensis | Leptospira noguchii | 99.1 | yes
SAMEA5168132 | Leptospira biflexa | Leptospira sp. | 21 |  
SAMEA5168202 | Leptospira noguchii | Leptospira alstonii | 97.4 | yes
SAMEA5168214 | Leptospira biflexa | Leptospira sp. | 95 | yes
SAMEA5168239 | Leptospira kirschneri | Leptospira sp. | 42 |  
SAMEA5168242 | Leptospira noguchii | Leptospira alstonii | 98.9 | yes
SAMEA5168266 | Leptospira borgpetersenii | Leptospira noguchii | 98.4 | yes
SAMEA5168277 | Leptospira sp. | Leptospira alstonii | 16.7 |  
SAMEA5168280 | Leptospira sp. | Leptospira sp. | 57.9 |  
SAMEA5168281 | Leptospira sp. | Leptospira alstonii | 14 |  
SAMEA5168316 | Leptospira meyeri | Leptospira sp. | 1.1 |  
SAMEA5168337 | Leptospira kirschneri | Leptospira borgpetersenii | 98.8 | yes


## Rest Low Genome Fraction Isolates


rest | reported species | NCBI blast analysis | percentage | keep?
-- | -- | -- | -- | --
SAMEA864091 | Leptospira interrogans | Leptospira sp. | 41.66 |  
SAMEA864145 | Leptospira interrogans | Leptospira sp. | 90.8 | yes
SAMEA864160 | Leptospira interrogans | Leptospira sp. | 97.1 | yes
SAMEA864166 | Leptospira interrogans | Leptospira weilii | 89.6 | yes
SAMN01162015 | Leptospira interrogans | Leptospira sp. | 99.6 | yes
SAMN01162028 | Leptospira interrogans | Leptospira sp. | 99.3 | yes
SAMN01919665 | Leptospira weilii | Leptospira weilii | 96.9 | yes

file for the two tables above:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/picardeau(rest)/picardeau(rest)_less_than_10_mapping.xlsx
```

## If the isolates were only classified as genus _Leptospira_, what is their identity of species?
- need to keep an eye on these isolates in downstream analysis

**Comment by rx32940** [comment]
- all isolates found with low genome fraction when mapped to their labeled species' reference genome were re-quasted with the reference genome of the species identified by NCBI STAT tool. 
[code for requast picardeau](https://github.com/rx32940/Lepto-Phylogeography/blob/9ff362356f86607fcd40e3066c57d2e8bb41c452/picardeau_assemble_pipeline.sh)
[code for requast rest](https://github.com/rx32940/Lepto-Phylogeography/blob/9ff362356f86607fcd40e3066c57d2e8bb41c452/rest_assembly_pipeline.sh)
report found in Dropbox: ```low_map_requast_picardeau(rest).html```
- all isolates when quasted with NCBI STAT identified species' reference genome have genome fraction over 70%.

- **SAMN01919665**, the isolate was labeled as _Leptospira weilii_ and has a strong signal wwhen analyzed by STAT for _Leptospira weilii serovar Ranarum str. ICFT_ shows very low generic fraction when mapped to _Leptospira weilii_ genome fraction, thus will be dropped from tthe list of assemblies.

**Comment by lsalvador** [comment]
what about the isolate SAMEA5168090. Why does it have such a N50 value?

**Comment by rx32940** [comment]
@lsalvador 
some of the Picardeau isolates have N50 much higher than others. For example, SAMEA5168092 has N50 as high as 1458.3 Kbps while most of the other Picardeau isolates only have N50 at ~50 Kbp range. I believe this difference could be due to the quality of the sequences. I will have an N50 distribution plot for all isolates up in the issue soon.

- [ ] isolate N50 density distribution plot

**Comment by rx32940** [comment]
### BWA MEM and Qualimap For NGS depth coverage

- to filter scaffolds with the best qualities, all **trimmed fastq** files for each isolate were mapped to their **corresponding reference genome** with their BAM file evaluated by **Qualimap**.
   - only scaffolds with **median coverage depth** higher than 30X and with at least 50% of the genome with higher than 30X of the coverage will be kept for further analysis.
[BWA MEM/ Qualimap Script for picardeau_313](https://github.com/rx32940/Lepto-Phylogeography/blob/8e9173f713a74266bd21c9e19ee502165c487584/picardeau_assemble_pipeline.sh)
[BWA MEM/ Qualimap Script for Rest_211](https://github.com/rx32940/Lepto-Phylogeography/blob/8e9173f713a74266bd21c9e19ee502165c487584/rest_assembly_pipeline.sh)
[BWA MEM/ Qualimap Script for dated_32](https://github.com/rx32940/Lepto-Phylogeography/blob/2e0194f6cce190f220abe6410902f98bd220cd10/dated_assemblies_pipeline.sh)

**Comment by rx32940** [comment]
## Assemblies Filtered and Selected 
- To Filter and selected **NCBI downloaded** assemblies: [assemblies_quality.ipynb](https://github.com/rx32940/Lepto-Phylogeography/blob/253fa4c9fc7141018564e80c1d9831b0384fb4d3/assemblies_quality.ipynb)
    - Coverage >= 30X
    - N50 >= 10,000
    - num of contigs <= 500

- To filter all **self_assembled** scaffolds: [self_assembled_qualities.ipynb](https://github.com/rx32940/Lepto-Phylogeography/blob/253fa4c9fc7141018564e80c1d9831b0384fb4d3/self_assembled_qualities.ipynb)
    - Qualimap reported median coverage >= 30X
    - at least 50% of the scaffold is >= 30X
    - N50 >= 10,000
    - num of contigs <= 500
    
- To summarize the qualities of all selected assemblies: [all_assemblies_quality.ipynb](https://github.com/rx32940/Lepto-Phylogeography/blob/dc090b055f20376f6aa6f928c880e9c12f4be9dc/all_assemblies_quality.ipynb)

** All the scripts and filtering standard listed in this thread is not the final version, need to make corrections correspondingly when BWA MEM mapping and Qualimap jobs finish running.

**Comment by rx32940** [comment]
## Summary of Assemblies' Quality
---
### NCBI Downloaded Assemblies

N50 Distribution
![Screen Shot 2020-03-22 at 2 37 51 PM](https://user-images.githubusercontent.com/31255012/77257488-01fe2000-6c4b-11ea-8323-3f0d39e56db5.png)

Total Length Distribution
![Screen Shot 2020-03-22 at 3 03 04 PM](https://user-images.githubusercontent.com/31255012/77258017-3d4e1e00-6c4e-11ea-87d9-c6b1bcccbc1e.png)

Number of Contigs Distribution
![Screen Shot 2020-03-22 at 3 48 02 PM](https://user-images.githubusercontent.com/31255012/77258974-8bfeb680-6c54-11ea-85e1-05fca644ea22.png)


### Self Assembled Assemblies

N50 Distribution

![Screen Shot 2020-03-25 at 2 44 59 PM](https://user-images.githubusercontent.com/31255012/77573554-3a019f00-6ea7-11ea-89b9-c62fc0512a42.png)

Total Length
![Screen Shot 2020-03-25 at 2 47 21 PM](https://user-images.githubusercontent.com/31255012/77573793-9238a100-6ea7-11ea-8bc7-363a578e5809.png)


Number of Contigs
![Screen Shot 2020-03-25 at 2 46 35 PM](https://user-images.githubusercontent.com/31255012/77573696-6e755b00-6ea7-11ea-9743-826dfc6bb1c9.png)




** Again this is not the final version of the quality summary report, will update once the coverage analysis is done

**Comment by rx32940** [comment]
Final **self-assembled assemblies** list stored in:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/all_self_assemblies_filtered.csv
```
Final **NCBI downloaded assemblies** list stored in:
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/ncbi_assemblies/ncbi_assemblies_filtered.csv
```

**Filters**:
1) coverage >= 20
2) num of contigs <= 500
3) N50 > = 10,000

For the self-assembled assemblies:
   - % Genome Fraction > 70% in QUAST report

Assemblies stores in:
```
/scratch/rx32940/Lepto_Work/(dated_32/ncbi_651/picardeau_313/rest_211)/assemblies
```
Also in ```\work\lslab``` for permanent storage:
```

```

**Comment by rx32940** [comment]
## Core Gene Builder Pipeline
- all 973 assemblies (including NCBI downloaded and self-assembled) were annotated by **Prokka** and use **Roary** to identify the core genes.
```
/mnt/c/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/ncbi+self_973.csv
```
- Prokka was used to create GFF3 format annotation files for each assembly.
script for the pipeline: [core_gene_builder.sh](https://github.com/rx32940/Lepto-Phylogeography/blob/13e331b29f8a51ddf14ed83c4a5300864f5e53a6/core_gene_builder.sh)
    - assign more memory when running the script. the job takes a large ram

### Analysis results

The analysis exited early because:
```
Number of clusters (242160) exceeds limit (50000). Multifastas not created. Please check the
spreadsheet for contamination from different species or increase the --group_limit parameter.
2020/03/30 07:59:50 Exiting early because number of clusters is too high
```


- In addition, the analysis also ended with **no core genes identified** (including softcore genes):

Core genes	(99% <= strains <= 100%)	0
Soft core genes	(95% <= strains < 99%)	0
Shell genes	(15% <= strains < 95%)	5307
Cloud genes	(0% <= strains < 15%)	236852
Total genes	(0% <= strains <= 100%)	242159

I found the answer to the situation on [Roary's website](https://sanger-pathogens.github.io/Roary/) in the FAQ's:
![Screen Shot 2020-03-30 at 2 32 02 PM](https://user-images.githubusercontent.com/31255012/77948326-4153ee80-7293-11ea-8541-a8826fd4eb60.png)

![Screen Shot 2020-03-30 at 2 33 56 PM](https://user-images.githubusercontent.com/31255012/77948494-7f511280-7293-11ea-8c1d-d55e462afead.png)

- many of my input assemblies are fragmented even after I limited the number of contigs to < 500. 
- although Leptospira has a relatively plastic genome, the resulting number of clusters is still very high comparing to the default cluster number and E. coli.

Next step:

- [x] to assess the quality of the input assemblies, I will run the Roary again with QC by Kraken2
like it was suggested above.
- [x] pick representative isolates from each species  for pan-genome analysis
    - manually pick to ensure good quality assemblies from each species were picked at a rational distributon
    - Xu et al 2016 picked 18/102
    - Guglielmini et al 2019 103/509

**Comment by rx32940** [comment]
- selected one isolate per species, excluding isolates with undetermined species. 
- 68 isolates were picked.
```
/Users/rx32940/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/one_isolate_per_species_68.csv
```

- [ ] run roary with 68 isolates
- [ ] run iqtree with 68 isolates
